python/pip_install/pip_repository.bzl

```diff
@@ -331,6 +337,7 @@ def _pip_repository_impl(rctx):
             "@{}//{}".format(rctx.attr.name, p) if rctx.attr.incompatible_generate_aliases else "@{}_{}//:pkg".format(rctx.attr.name, p)
```

Comment:
Shouldn't we remove the files that are in `groups` from the `requirements` array? This would prevent users getting errors when generating the `gazelle` manifest.

Reply:
We should be. This is an oversight on my part.

Reply:
So something's not quite right over there. We don't need to filter this list, because we're using the same alias machinery for the `:whl` targets that we are for the `:pkg` targets. So to take your sphinx example, you get `@pip_sphinx//:whl` which an alias for `@pip__groups//:sphinx_whl`, which wraps up all the `:_whl` filegroups which lack inter-group dependencies. That this list is taking multiple dependencies (via aliases) on the group rules works out just fine.

Using that machinery --

```shellsession
$ bazel cquery --output=files 'deps(@pip//sphinx:whl)'
INFO: Analyzed target @pip//sphinx:whl (1 packages loaded, 1 target configured).
INFO: Found 1 target...
external/pip_sphinxcontrib_applehelp/sphinxcontrib_applehelp-1.0.7-py3-none-any.whl
external/pip_sphinx/sphinx-7.2.6-py3-none-any.whl
external/pip_sphinxcontrib_devhelp/sphinxcontrib_devhelp-1.0.5-py3-none-any.whl
external/pip_sphinxcontrib_serializinghtml/sphinxcontrib_serializinghtml-1.1.9-py3-none-any.whl
external/pip_sphinxcontrib_htmlhelp/sphinxcontrib_htmlhelp-2.0.4-py3-none-any.whl
external/pip_sphinxcontrib_qthelp/sphinxcontrib_qthelp-1.0.6-py3-none-any.whl
external/pip_packaging/packaging-23.2-py3-none-any.whl
external/pip_importlib_metadata/importlib_metadata-5.2.0-py3-none-any.whl
external/pip_snowballstemmer/snowballstemmer-2.2.0-py2.py3-none-any.whl
external/pip_alabaster/alabaster-0.7.13-py3-none-any.whl
external/pip_requests/requests-2.31.0-py3-none-any.whl
external/pip_pygments/Pygments-2.16.1-py3-none-any.whl
external/pip_sphinxcontrib_jsmath/sphinxcontrib_jsmath-1.0.1-py2.py3-none-any.whl
external/pip_docutils/docutils-0.20.1-py3-none-any.whl
external/pip_jinja2/Jinja2-3.1.2-py3-none-any.whl
external/pip_imagesize/imagesize-1.4.1-py2.py3-none-any.whl
external/pip_babel/Babel-2.13.1-py3-none-any.whl
external/pip_idna/idna-3.4-py3-none-any.whl
external/pip_markupsafe/MarkupSafe-2.1.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
external/pip_charset_normalizer/charset_normalizer-3.3.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
external/pip_zipp/zipp-3.11.0-py3-none-any.whl
external/pip_urllib3/urllib3-2.0.7-py3-none-any.whl
external/pip_certifi/certifi-2023.7.22-py3-none-any.whl
INFO: Elapsed time: 0.252s, Critical Path: 0.00s
INFO: 0 processes.
INFO: Build completed successfully, 0 total actions
```

https://gist.github.com/arrdem/5878f481c86b046df339af494bd1be7b

---

The bad news is that for reasons that aren't obvious to me the `//:gazelle_python_manifest.update` rule isn't picking up all the component wheels for some reason despite `bazel query` giving the desired results. Do we want to be referencing the concrete `:_whl` targets here if the Gazelle indexing machinery doesn't compute a closure and traverse filegroup deps fully?

Reply:
Ah. The manifest generator rule doesn't traverse `data=[]` in filegroups. Easy fix.

Reply:
Gazelle manifest generation works fine now